### PR TITLE
WIP: feat: add support for initContainer in dev mode

### LINKED ISSUE #2360: `devspace dev` for initContainers
**Is your feature request related to a problem?**
Yes, my problem is that i need to debug a init container. This isn't possible right now as simply can't select the init container.
My current kind of workaround is to temporary modify the deployment and move the "initContainer" to "containers". But this is not a great solution, because i have to modify the deployments :-1:

**Which solution do you suggest?**
Allowing the user to select if the dev section is for containers or init containers.

**Which alternative solutions exist?**
None that i know of.

**Additional context**

```
$ devspace --version               
devspace version 6.1.1
```

/kind feature

### PR DESCRIPTION
**Pull request by pratikjagrut**
**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind feature

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves #2360 

**Please provide a short message that should be published in the DevSpace release notes**
Fixed an issue where DevSpace ...

**What else do we need to know?**

### CODE REVIEW

File: pkg/devspace/devpod/devpod.go [hunk 1]
```diff
@@ -315,9 +323,16 @@ func (d *devPod) start(ctx devspacecontext.Context, devPodConfig *latest.DevPod,
...
+	if !devPodConfig.InitContainer {
+		err = d.startServices(ctx, devPodConfig, newTargetSelector(selectedPod.Pod.Name, selectedPod.Pod.Namespace, selectedPod.Container.Name, parent), opts, parent)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = d.startServices(ctx, devPodConfig, targetselector.NewTargetSelector(options), opts, parent)
+		if err != nil {
+			return err
+		}
```
Comment (lizardruss) [inline pkg/devspace/devpod/devpod.go:335]: Maybe reduce duplication by extracting a `targetSelector` var.

File: pkg/devspace/devpod/devpod.go [hunk 2]
```diff
@@ -268,9 +274,11 @@ func (d *devPod) start(ctx devspacecontext.Context, devPodConfig *latest.DevPod,
...
 			}
 		}
-		for _, container := range selectedPod.Pod.Spec.InitContainers {
-			if container.Name == devContainer.Container {
-				return true
+		if devPodConfig.DevContainer.InitContainer {
+			for _, container := range selectedPod.Pod.Spec.InitContainers {
+				if container.Name == devContainer.Container {
+					return true
+				}
```
Comment (lizardruss) [inline pkg/devspace/devpod/devpod.go:281]: Could reduce duplication by extracting a `containers` var

File: pkg/devspace/devpod/devpod.go [hunk 3]
```diff
@@ -247,9 +248,14 @@ func (d *devPod) start(ctx devspacecontext.Context, devPodConfig *latest.DevPod,
...
-		WithWaitingStrategy(targetselector.NewUntilNewestRunningWaitingStrategy(time.Millisecond * 500)).
-		WithSkipInitContainers(true)
+		ApplyConfigParameter("", devPodConfig.LabelSelector, imageSelector, devPodConfig.Namespace, "")
+
+	if devPodConfig.InitContainer {
+		options = options.WithWaitingStrategy(targetselector.NewUntilInitContainerIsRunningWaitingStrategy(time.Millisecond*500, devPodConfig))
+	} else {
+		options = options.WithWaitingStrategy(targetselector.NewUntilNewestRunningWaitingStrategy(time.Millisecond * 500))
+	}
+
```
Comment (lizardruss) [inline pkg/devspace/devpod/devpod.go:258]: Was it not possible to modify the existing waiting strategy?

File: pkg/devspace/services/targetselector/until_init_container_running.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	if selector.IsPodTerminating(container.Pod) {
+		return false
+	}
+	for _, cs := range container.Pod.Status.InitContainerStatuses {
+		if cs.Name == container.Container.Name && !cs.Ready && cs.State.Running != nil {
+			return true
+		}
+	}
+	return false
+}
```
Comment (lizardruss) [inline pkg/devspace/services/targetselector/until_init_container_running.go:120]: The other implementation reports pod problems. What does the behavior of this have when there is an image pull backoff for the init container, or if it's in a crash loop?

**Review by lizardruss — CHANGES_REQUESTED** [review]
In general this looks good. I would like to reduce code duplication, if it's possible without hurting readability.

Also some testing, either unit or e2e. At least one happy path where everything works as expected, and one where there's issues with the init container.

**Comment by alexandrem** [comment]
Is there anyone still working on this?

This is a blocking issue for me and I'm interested in this fix.

**Comment by alexandrem** [comment]
Is there anyone still working on this?

This is a blocking issue for me and I'm interested in this fix.

**Comment by FabianKramm** [comment]
@alexandrem since this is impacting a lot of areas in DevSpace, we have postponed the change

**Pull request closed without merging**